libvast/builtins/operators/tail.cpp

```diff
@@ -0,0 +1,115 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/concept/parseable/numeric/integral.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/legacy_pipeline.hpp>
+#include <vast/logger.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/table_slice.hpp>
+
+#include <arrow/type.h>
+
+namespace vast::plugins::tail {
+
+namespace {
+
+class tail_operator final : public crtp_operator<tail_operator> {
+public:
+  explicit tail_operator(uint64_t limit) : limit_{limit} {
+  }
+
+  auto operator()(generator<table_slice> input) const
+    -> generator<table_slice> {
+    auto remaining = limit_;
+    for (auto&& slice : input) {
+      slice = vast::tail(slice, remaining);
+      VAST_ASSERT(remaining >= slice.rows());
+      remaining -= slice.rows();
+      co_yield std::move(slice);
+      if (remaining == 0) {
+        break;
+      }
+    }
```

Comment:
As far as I can tell, this does not have the semantics that we want. For example, if we receive 3 tables slices of 10 events each here, and we have `tail 15`, I would expect the last 5 events of the second `table_slice`, and all of the events of the last `table_slice` (so the events 15-30), while this here would output the first 10 events, followed by events 15-20, right? I think we need some kind of buffer.

Reply:
Yep, indeed.

Reply:
We also need the buffer because the result of tail could be heterogeneous. Imagine the following scneario:

1. slice with 15 rows of schema A
2. slice with 1 row of schema B
3. slice with 4 rows of schema A

Then your output for `tail 10` should be (in order) 

1. the last 5 rows of the first slice
2. all 1 rows of the second slice
3. all 4 rows of the last slice

Reply:
Luckily, I already had a buffer in my `tail` `legacy_pipeline_operator` implementation.

I implemented a buffer that is order-sensitive.